### PR TITLE
ALD-11 | Add reCAPTCHA component

### PR DESCRIPTION
**Pull request by NuriaExtremadouro**
## Description

- Added reCAPTCHA component.
- Added tests and docs in the Readme. Note that the test for the component is rather simple because we would just testing a third-party library.

## Related Issue

- [ALD-11](https://csmdigital.atlassian.net/browse/ALD-11) -> Adding the reCAPTCHA component.
- [ALD-12](https://csmdigital.atlassian.net/browse/ALD-12) -> Receiving the key from the form data passed to the library.

## How Has This Been Tested?

Tested both manually and with unit tests to verify that:
- The form submits when there's no reCAPTCHA.
- The form submits when there's reCAPTCHA but it doesn't detect a bot.
- The reCAPTCHA modal shows to verify the user is not a bot when it detects bot-like behavior. Upon completing the reCAPTCHA (selecting images that match a word) the form submits normally.

Note that at the end of the testing the reCAPTCHA question has been removed from the example form. The reason is because it requires a key to work, so it would block anyone from using the example since they'd need a valid reCAPTCHA key to run it locally.

## Screenshots (if appropriate):

ReCAPTCHA presence in the page:
Captura de Pantalla 2024-02-28 a las 13 55 48

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/Questions/Recaptcha/__tests__/recaptcha.test.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  const { getByTestId } = render(
+    <QuestionRecaptcha
+      recaptchaRef={recaptchaRef}
+      formDataValues={formDataValues}
+      onSubmitForm={onSubmitForm}
+      question={question}
+    />
+  )
+
+  expect(getByTestId('recaptcha-test')).toBeTruthy()
```
Comment (matyasjay) [inline src/Questions/Recaptcha/__tests__/recaptcha.test.js:26]: Wondering if we can make the program verify the captcha 🤣

Comment (NuriaExtremadouro) [reply src/Questions/Recaptcha/__tests__/recaptcha.test.js:26]: To be fair that would disprove its ability to block bots hahaha

File: src/Questions/Recaptcha/index.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/** @jsx jsx */
+import { jsx } from 'theme-ui'
+
+const QuestionRecaptcha = ({ recaptchaRef, formDataValues, onSubmitForm, question }) => {
+  const onReCAPTCHAChange = async (captchaCode) => {
+    if (!captchaCode) {
+      return
+    }
+
+    recaptchaRef.current.reset()
```
Comment (matyasjay) [inline src/Questions/Recaptcha/index.js:13]: I thing React refs are nullable. Might need to add optional chaining here.
```suggestion
    recaptchaRef.current?.reset()
```

File: src/builder.js [hunk 3]
```diff
@@ -42,6 +43,9 @@ const FormBuilder = ({
   ...props
 }) => {
   const useFormObj = useForm({ defaultValues: { formatDate: '' } })
+  const [formDataValues, setFormDataValues] = React.useState({})
+  const hasRecaptcha = form.questions.some(question => question.type === 'recaptcha')
+  const recaptchaRef = React.createRef(null)
```
Comment (matyasjay) [inline src/builder.js:48]:
```suggestion
  const recaptchaRef = useRef(null)
```
Just dont forget to import the hook first.

File: src/builder.js [hunk 4]
```diff
@@ -299,7 +311,13 @@ const FormBuilder = ({
 
   const onSubmit = async (data) => {
     if (isLoading) return
-    onSubmitForm(await formatData(data))
+    if (hasRecaptcha) {
+      recaptchaRef.current.execute()
```
Comment (matyasjay) [inline src/builder.js:315]: same as above, optional chain needed

File: README.md [hunk 5]
```diff
@@ -778,6 +778,26 @@ Autocomplete basic example
...
+|---	|---	|:---:	|:---:	|
+|   name*	|  ReCAPTCHA name  	|  string 	|  - 	|
+|   type*   | Must be `recaptcha`| string | - |
+|   recaptchaKey*	|  Key to make ReCAPTCHA work. Failing to provide this will lead to the ReCAPTCHA element showing with an error text inside. 	|  string  	|   -	|
+
+This feature is implemented using `react-google-recaptcha`. By default it shows at the bottom-right of your app and will only show the ReCAPTCHA modal when bot-like behavior is detected filling the form.
+
+#### ReCAPTCHA example
+
+```yaml
```
Comment (inigomarquinez) [inline README.md:793]:
```suggestion
```json
```

**Review by matyasjay — CHANGES_REQUESTED** [review]
fyi @NuriaExtremadouro , the code looks great! 
I've added a few comments about patterns, but nothing big, cheers!

**Review by inigomarquinez — CHANGES_REQUESTED** [review]
LGTM!

Just one small typo in the docs

File: README.md (hunk 5 above)
Comment (inigomarquinez) [reply README.md]: Don't worry about them. Change them if you want but nothing to worry about

**Review by matyasjay — APPROVED** [review]
Brilliant! Thanks! ❤️

**Review by inigomarquinez — APPROVED** [review]
Awesome Nuria!

**Comment by inigomarquinez** [comment]
@SocketSecurity ignore npm/core-js-pure@3.35.1
@SocketSecurity ignore npm/core-js@3.35.1

**Comment by inigomarquinez** [comment]
@all-contributors please add @NuriaExtremadouro for code,doc

**Pull request merged**